R/doeResponseSurfaceMethodology.R

```diff
@@ -787,26 +797,112 @@ doeResponseSurfaceMethodology <- function(jaspResults, dataset, options, ...){
-  for (i in 1:(ncol(data)-1)) {
-    unique_l[[i]] <- unique(data[,i])
+  desire_list <- list()
+  k <- 1
+  if (length(options[["rsmMin"]]) > 0) {
+    for (i in 1:length(options[["rsmMin"]])) {
+      desire_list[[k]] <- desirability::dMin(options[["rsmMin"]][[i]][[2]], options[["rsmMin"]][[i]][[1]])
+      k <- k +1
+    }
+  }
+  if (length(options[["rsmMax"]]) > 0) {
+    for (i in 1:length(options[["rsmMax"]])) {
+      desire_list[[k]] <- desirability::dMin(options[["rsmMax"]][[i]][[2]], options[["rsmMax"]][[i]][[1]])
+      k <- k +1
+    }
+  }
+  if (length(options[["rsmTar"]]) > 0) {
+    for (i in 1:length(options[["rsmTar"]])) {
+      desire_list[[k]] <- desirability::dMin(options[["rsmTar"]][[i]][[2]],options[["rsmTar"]][[i]][[3]], options[["rsmTar"]][[i]][[1]])
+      k <- k +1
+    }
+  }
+
+  desire_final <- do.call(desirability::dOverall, desire_list)
+
+  if (length(options[["rsmMax"]]) + length(options[["rsmMin"]]) + length(options[["rsmTar"]]) == 1)
+    desire_final <- desire_final[[1]][[1]]
+
+  op  <- data.frame(x1 = seq(-1,1,0.25))
+  op1 <- length(options[["rsmVariables"]])
+
+  if (op1 > 1) {
+    for (i in 2:op1) {
+      op <- cbind(op, seq(-1,1,0.25))
+    }
+    names(op) <- paste("x",1:op1, sep = "")
   }
-  # print(unique_l)
-  comb <- expand.grid(unique_l)
-  # print(names(data))
-  unique_l2 <- list()
-  for (i in 1:(ncol(data)-1)) {
-    unique_l2[[i]] <- unique(data[,i])
+
+
+  you <- expand.grid(op)
+
+  # print(desire_final)
+  # print(length(desire_final))
+  # print(predict(rsm, newdata = you[i,]))
+  # print(desire_final)
+  print(you[i,])
+  rsmOpt_2 <- function(you_2, rsm, dObject, space = "square") {
+    conv <- numeric()
+    for (j in 1:length(rsm)) {
+      conv[j] <- predict(rsm[[j]], newdata = you[i,])
+    }
+    out <- do.call(predict, list(object = dObject, newobject = conv))
+
+    if (space == "circular") {
+      if (sqrt(sum(you[i,]^2)) > 1.682)
+        out <- 0
+    }
+    else if (space == "square")
+      if (any(abs(you[i,]) > 1.682))
+        out <- 0
+    out
   }
-  for (i in 1:length(unique_l)) {
-    for (j in 1:length(unique_l[[i]])) {
-      unique_l2[[i]][j] <- rsm::val2code(eval(parse(text = paste("c(",names(data)[i], "=", as.character(unique_l[[i]][j]),")", sep = ""))), codings = rsm::codings(rsm))
+
+  tmp <- list()
+  for (i in 1:dim(you)[1]) {
+    tmp[[i]] <- optim(as.vector(you[i,]), rsmOpt_2, dObject = desire_final,
+                      rsm = rsm, space = "square", control = list(fnscale = -1))
+    if (i == 1) {
+      best_2 <- tmp[[i]]
+    }else {
+      if (tmp[[i]][["value"]] > best_2[["value"]])
+        best_2 <- tmp[[i]]
     }
   }
-  comb2 <- expand.grid(unique_l2)
-  # print(comb)
-  # print(comb2)
 
-  coef <- summary(rsm)[[4]][,1]
+  print(tmp)
+
+  desirability_table <- createJaspContainer()
+  jaspResults[["desirability_table"]] <- desirability_table
+  desirability_table$dependOn(options = c("rsmMin", "rsmMax", "rsmVariables", "rsmTar"))
```

Comment:
right now, you're recreating this container (and all its contents!) every time no matter the dependencies.

```suggestion
  if (is.null(jaspResults[["desirability_table"]])) {
    desirability_table <- createJaspContainer()
    jaspResults[["desirability_table"]] <- desirability_table
    desirability_table$dependOn(options = c("rsmMin", "rsmMax", "rsmVariables", "rsmTar"))
  } else {
     desirability_table <- jaspResults[["desirability_table"]]
  }
```